### DISCUSSION TITLE
Данные о прочитанности комментариев

### DISCUSSION THREAD

**Discussion by AndreyKostik** [totumonline/totum-ru-issues-and-discussions#74] [Как это сделать?] [answered]
Добрый день.
По структуре работы возникла необходимость сообщать пользователям на email о наличии непрочитанных комментариев. Соответственно, возник вопрос, как определять наличие таких комментариев.
Первый подход к проблеме был таков:
- получаем общее количество комментариев из поля через sеlect;
- получаем количество прочитанных сообщений из _comments_viewed;
- сравниваем и на основании этих данных действуем.

Проблема возникла на втором пункте, тк totum не дает прямого доступа к _comments_viewed ("таблица не найдена", что, в целом, логично).
Вопросы:
Существует ли способ обращения с помощью кода totum к _comments_viewed без создания отдельных скриптов?
Есть ли более очевидный способ получения данных о прочитанности комментариев?

**Comment by ttm-alexeypolunin** [comment]
Я понял задачу, мы добавим в .38 версию функцию типа `select`, которая будет ожидать координаты поля и `list` из `id` пользователей (ну или `list` из одного `id`).

Возвращать будет `rowList` содержащий ключи `users`, `num` и `comments`.

Назовем наверное `selectUnreadComments`

Будет на сл неделе тк .39 сейчас стейджится.

**Comment by ttm-alexeypolunin** [comment, marked as answer]
Обновите — сегодня в [.38 +] закинули и проверили работоспособность.

```
= : selectUnreadComments(users: json`[1,3,4]`; table: $#ntn; field: 'komments'; id: 5)
```

`users` — ожидает список id пользователей

`field` — поле типа комментарий.

`id` — id строки если коммент в строчной части

Так же как и у `select` у этой функции есть `cycle` если нужно брать из цикла.

В результате получается вот такой массив данных:

<img width="887" alt="Снимок экрана 2022-02-21 в 15 30 04" src="https://user-images.githubusercontent.com/55755565/154963233-ac499bb8-857e-4c1e-b684-83dde4a50b4a.png">

В нем на каждого запрошенного пользователя будет строка. Пользователь обозначен в `user`, количество непрочитанных комментов в `num`. Сами комменты (непрочитанные) в `comments`.

Попробуйте — если чего лажает пишите, исправим.

**Reply by AndreyKostik** [reply]
Вот такая конструкция в строчной части:

```
= : selectUnreadComments (users: $list; table: $#ntn; field: #kommentarii; id: #id)
list: listCreate(item: 1)
```

Вызывает сбой всей таблицы:
Fatal error: Uncaught TypeError: array_key_exists(): Argument #1 ($key) must be a valid array offset type in.....   /totum/tableTypes/aTable.php:1187

**Reply by ttm-alexeypolunin** [reply]
У вас name поля из которого надо получить данные о непрочитанности хранится в поле `kommentarii`?

Скорее всего `kommentarii` это и есть поле с комментариями, тогда:

```
= : selectUnreadComments (users: $list; table: $#ntn; field: 'kommentarii'; id: #id)
list: listCreate(item: 1)
```

Чтоб он ругался нормально мы сделаем на этой неделе.

**Reply by ttm-alexeypolunin** [reply]
И еще важно, что бы вы были обновлены до **сегодняшнего** коммита!

**Reply by AndreyKostik** [reply]
Точно. Спасибо, что указали на ошибку!

**Comment by AndreyKostik** [comment]
Добрый день. Не очень понятна причина такого поведения функции:
В строчной части есть поле `kommentarii`, и рядом положил колонку с `selectUnreadComments`, для проверки и настройки.

Функция, как писал выше, выглядит так:

```
= : selectUnreadComments(users: $list; table: $#ntn; field: 'kommentarii'; id: #id)
list: listCreate(item: 1)
```
Оставляю в `kommentarii` два комментария от другого пользователя, возвращаюсь под пользователем 1, и вижу в `selectUnreadComments`вот так:

![image](https://user-images.githubusercontent.com/99019247/156367158-beb4eef1-8326-4f8c-a6d7-920067e582d1.png)

То есть три комментария вместо двух. Самый первый – ранее уже прочитанный пользователем 1 старый комментарий. Соответственно, 1 и 2 - новые.
Далее прочитываю комментарии путем открывания `kommentarii` (под пользователем 1). Подчеркивание снимается. Захожу в  `selectUnreadComments` и вижу вот так:

![image](https://user-images.githubusercontent.com/99019247/156367333-ff5aed8c-a0ee-4c48-8f2d-5425388058bc.png)

Остался самый последний комментарий. 
По идее, здесь в этот момент не должно же быть записей?

**Reply by ttm-alexeypolunin** [reply]
Я думаю вы положили функцию в соседнее поле в Код — а он вычисляется при изменении в единице пересчета. Когда вы открываете комментарии не происходит пересчета строки. Поэтому данные в соседнем поле остаются без изменения.

**Reply by AndreyKostik** [reply]
Пересчет происходит.
Было два непрочитанных комментария - ячейка показывает, что их три.
Как только прочитываешь комментарии - происходит пересчет и ячейка показывает, что есть один непрочитанный.
![image](https://user-images.githubusercontent.com/99019247/156729387-2b9a5a75-0ed5-4b71-9a07-38132f2f4db7.png)

**Reply by ttm-alexeypolunin** [reply]
Да, идет отставание на 1 шаг. Если еще раз пересчитать строку то количество должно выровняться как надо.

Те когда в той же строке мы вызываем `...unread...` у БД еще нет информации о прочитанности тк она добавляется позже. Попробуем это пофиксить.

**Reply by AndreyKostik** [reply]
Тогда, по идее, поле должно обновиться при пересчете таблицы. В моем случае ни повторное открытие таблицы, ни `reCalculate` ничего не меняет. По-прежнему, якобы, один непрочитанный комментарий. Таблица - расчетная в дереве.

**Reply by ttm-alexeypolunin** [reply]
Ага, разраб заэкономил на спичках — ща будем исправлять.

По тому, что он показывает на шаг назад — это правильная логика поведения всех функций типа `select` (начинаю сам забывать некоторые штуки). Поэтому если хочется иметь поле в этой же строке, то надо будет дополнительно пересчитать по коду действия.

**Reply by AndreyKostik** [reply]
Ну и, стесняясь, прошу запланировать что-то типа `SetAsRead`. Было бы здорово.

**Reply by ttm-alexeypolunin** [reply]
На местере — обновите и попробуйте. В расчетной запаздывания не будет в простых будет.

`setAsRead` — записал, сделаем гденить на сл неделе

**Reply by AndreyKostik** [reply]
Обновил все, что есть. Сожалею, но поведение функции не поменялось.

**Reply by ttm-alexeypolunin** [reply]
Попытка 100500 — обновите
